docs/ldml/tr35-keyboards.md

```diff
@@ -1799,62 +2100,70 @@ Keyboarding applications typically work, but are not required to, in one of two
+    <transformGroup>
+        <transform from="\u0915\u094D\u0936"/>
+    </transformGroup>
 </transforms>
 ```
 
-Here there is no `@to` attribute since the whole string is being deleted. This is not uncommon in the backspace transforms.
+Here the optional attribute `@to` attribute is omitted, since the whole string is being deleted. This is not uncommon in the backspace transforms.
 
 A more complex example comes from a Burmese visually ordered keyboard:
 
 ```xml
 <transforms type="backspace">
-    <!-- Kinzi -->
-    <transform from="[\u1004\u101B\u105A]\u103A\u1039" />
+    <transformGroup>
+        <!-- Kinzi -->
+        <transform from="[\u1004\u101B\u105A]\u103A\u1039" />
 
-    <!-- subjoined consonant -->
-    <transform from="\u1039[\u1000-\u101C\u101E\u1020\u1021\u1050\u1051\u105A-\u105D]" />
+        <!-- subjoined consonant -->
+        <transform from="\u1039[\u1000-\u101C\u101E\u1020\u1021\u1050\u1051\u105A-\u105D]" />
 
-    <!-- tone mark -->
-    <transform from="\u102B\u103A" />
+        <!-- tone mark -->
+        <transform from="\u102B\u103A" />
 
-    <!-- Handle prebases -->
-    <!-- diacritics stored before e-vowel -->
-    <transform from="[\u103A-\u103F\u105E-\u1060\u1082]\u1031" to="\u1031" />
+        <!-- Handle prebases -->
+        <!-- diacritics stored before e-vowel -->
+        <transform from="[\u103A-\u103F\u105E-\u1060\u1082]\u1031" to="\u1031" />
 
-    <!-- diacritics stored before medial r -->
-    <transform from="[\u103A-\u103B\u105E-\u105F]\u103C" to="\u103C" />
+        <!-- diacritics stored before medial r -->
+        <transform from="[\u103A-\u103B\u105E-\u105F]\u103C" to="\u103C" />
 
-    <!-- subjoined consonant before e-vowel -->
-    <transform from="\u1039[\u1000-\u101C\u101E\u1020\u1021]\u1031" to="\u1031" />
+        <!-- subjoined consonant before e-vowel -->
+        <transform from="\u1039[\u1000-\u101C\u101E\u1020\u1021]\u1031" to="\u1031" />
 
-    <!-- base consonant before e-vowel -->
-    <transform from="[\u1000-\u102A\u103F-\u1049\u104E]\u1031" to="\uFDDF\u1031" />
+        <!-- base consonant before e-vowel -->
+        <transform from="[\u1000-\u102A\u103F-\u1049\u104E]\u1031" to="\uFDDF\u1031" />
 
-    <!-- subjoined consonant before medial r -->
-    <transform from="\u1039[\u1000-\u101C\u101E\u1020\u1021]\u103C" to="\u103C" />
+        <!-- subjoined consonant before medial r -->
+        <transform from="\u1039[\u1000-\u101C\u101E\u1020\u1021]\u103C" to="\u103C" />
 
-    <!-- base consonant before medial r -->
-    <transform from="[\u1000-\u102A\u103F-\u1049\u104E]\u103C" to="\uFDDF\u103C" />
+        <!-- base consonant before medial r -->
+        <transform from="[\u1000-\u102A\u103F-\u1049\u104E]\u103C" to="\uFDDF\u103C" />
 
-    <!-- delete lone medial r or e-vowel -->
-    <transform from="\uFDDF[\u1031\u103C]" />
+        <!-- delete lone medial r or e-vowel -->
+        <transform from="\uFDDF[\u1031\u103C]" />
+    </transformGroup>
 </transforms>
 ```
 
```

Comment:
```suggestion
If no specified transform among all `transformGroup`s under the `<transforms type="backspace">` element matches, a default will be used instead - an implied transform that simply deletes the most recent codepoint.
```

Or something similar, yet more accurate, in wording.  Just to be clear that keyboard authors don't have to directly handle the most simple & common cases.

The "implied transform", to my understanding, should match "standard"/common backspace behavior.  Or maybe it should be "a single character cluster"?  Either way, we want something that is simple and straightforward to understand and incorporate into a keyboard developer's mental model... and probably something that isn't reliant upon script-specific details.

----

Being explicit here may also be fairly important - I've seen certain platforms (iOS) auto-delete more than a single codepoint in certain cases.  For example, it may delete multiple codepoints when combining emoji are involved:

https://github.com/keymanapp/keyman/blob/05759dc95f0148edb6d47b00a3ff081536b06586/ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift#L341-L351

We don't want keyboard authors to have surprises in this regard, where one platform interprets the spec differently such that default backspace implementation differs per platform.

Other fun context regarding this:

https://github.com/keymanapp/keyman/pull/4427#issuecomment-777905602

In the comment linked above, iOS's default backspace (`.deleteBackward()`) behavior deleted _four_ codepoints in a single call!  To be clear, those four codepoints were a single Khmer character cluster - but that means the iOS default is to delete a _cluster_, not a _codepoint_.

It's possible that my assumption about the default being a single codepoint is wrong, and that it should be a character-cluster instead.  Either way, this is why I believe it needs explicit definition here - so that we don't leave vendors any "wiggle room" that would affect cross-platform behavior.

Reply:
Possibly of note - it is my understanding that "character cluster" boundaries, internally, could be inferred from the "reorder" mechanism?  Not that this is likely to be seen as sufficiently "explicit and clear" enough to be relied upon by your average keyboard author.

The issue I see with relying on "character cluster" default backspacing, if that would otherwise be our answer - what defines a "cluster" for minority language keyboards - and in a manner that can be incorporated into a compiled LDML keyboard?  I suppose script-specific knowledge may suffice - but I believe that the only mechanism currently in the keyboards with access to such knowledge would be the reorder rules.

Reply:
Per @mcdurdin - "deletes a single codepoint, unless it's an emoji cluster.  For emoji clusters, a backspace deletes one full combined-emoji cluster."

That's our take, anyway.

Reply:
Let's discuss in a few.

Reply:
I added a comment.  Does not address emoji yet but let's discuss.

Reply:
Think this is now covered, reopen if not

Reply:
LGTM